### PR TITLE
Upgrade Python 3.8 to 3.10

### PR DESCRIPTION
**Pull request by pamelafox**
Python 3.9 was released 2020-10-05 and has been supported by Azure App Service and Azure Functions for a while, so it should be fine to update the samples to all use 3.9. 

We could almost update them to 3.10, but we're waiting for that to be out of preview for Azure Functions:
https://learn.microsoft.com/en-us/azure/azure-functions/functions-reference-python?tabs=asgi%2Capplication-level&pivots=python-mode-configuration#python-version

The template-tests job should be kicked off for this to confirm no issues.

### CODE REVIEW

**Comment by jongio** [comment]
Can you see if you can find out when 3.10 will be out of preview?

**Review by vhvb1989 — APPROVED** [review]

File: templates/common/.devcontainer/devcontainer.json/python/devcontainer.json
Comment (pamelafox) [inline templates/common/.devcontainer/devcontainer.json/python/devcontainer.json:21]: Just changed this to 3.9 as well, for consistency.

**Comment by pamelafox** [comment]
The CI failed on files I did not change:

```
Run cspell lint '**/*' --config ./.vscode/cspell.yaml --relative --no-progress
./schemas/v1.0/azure.yaml.json:159:30 - Unknown word (postdeploy)
./schemas/v1.0/azure.yaml.json:16[4](https://github.com/Azure/azure-dev/actions/runs/4057804310/jobs/6983992673#step:5:5):30 - Unknown word (prerestore)
./schemas/v1.0/azure.yaml.json:169:30 - Unknown word (postrestore)
./schemas/v1.0/azure.yaml.json:179:30 - Unknown word (postpackage)
./schemas/v1.0/azure.yaml.json:2[5](https://github.com/Azure/azure-dev/actions/runs/4057804310/jobs/6983992673#step:5:6)3:18 - Unknown word (preprovision)
./schemas/v1.0/azure.yaml.json:258:18 - Unknown word (postprovision)
./schemas/v1.0/azure.yaml.json:2[6](https://github.com/Azure/azure-dev/actions/runs/4057804310/jobs/6983992673#step:5:7)3:18 - Unknown word (preinfracreate)
./schemas/v1.0/azure.yaml.json:268:18 - Unknown word (postinfracreate)
./schemas/v1.0/azure.yaml.json:2[7](https://github.com/Azure/azure-dev/actions/runs/4057804310/jobs/6983992673#step:5:8)3:1[8](https://github.com/Azure/azure-dev/actions/runs/4057804310/jobs/6983992673#step:5:9) - Unknown word (preinfradelete)
./schemas/v1.0/azure.yaml.json:278:18 - Unknown word (postinfradelete)
./schemas/v1.0/azure.yaml.json:283:18 - Unknown word (predown)
./schemas/v1.0/azure.yaml.json:288:18 - Unknown word (postdown)
./schemas/v1.0/azure.yaml.json:2[9](https://github.com/Azure/azure-dev/actions/runs/4057804310/jobs/6983992673#step:5:10)3:18 - Unknown word (preup)
./schemas/v1.0/azure.yaml.json:298:18 - Unknown word (postup)
./schemas/v1.0/azure.yaml.json:308:18 - Unknown word (postdeploy)
./schemas/v1.0/azure.yaml.json:3[13](https://github.com/Azure/azure-dev/actions/runs/4057804310/jobs/6983992673#step:5:14):[18](https://github.com/Azure/azure-dev/actions/runs/4057804310/jobs/6983992673#step:5:19) - Unknown word (prerestore)
./schemas/v1.0/azure.yaml.json:318:18 - Unknown word (postrestore)
```

**Comment by rajeshkamal5050** [comment]
> The CI failed on files I did not change:
> 
> ```
> Run cspell lint '**/*' --config ./.vscode/cspell.yaml --relative --no-progress
> ./schemas/v1.0/azure.yaml.json:159:30 - Unknown word (postdeploy)
> ./schemas/v1.0/azure.yaml.json:16[4](https://github.com/Azure/azure-dev/actions/runs/4057804310/jobs/6983992673#step:5:5):30 - Unknown word (prerestore)
> ./schemas/v1.0/azure.yaml.json:169:30 - Unknown word (postrestore)
> ./schemas/v1.0/azure.yaml.json:179:30 - Unknown word (postpackage)
> ./schemas/v1.0/azure.yaml.json:2[5](https://github.com/Azure/azure-dev/actions/runs/4057804310/jobs/6983992673#step:5:6)3:18 - Unknown word (preprovision)
> ./schemas/v1.0/azure.yaml.json:258:18 - Unknown word (postprovision)
> ./schemas/v1.0/azure.yaml.json:2[6](https://github.com/Azure/azure-dev/actions/runs/4057804310/jobs/6983992673#step:5:7)3:18 - Unknown word (preinfracreate)
> ./schemas/v1.0/azure.yaml.json:268:18 - Unknown word (postinfracreate)
> ./schemas/v1.0/azure.yaml.json:2[7](https://github.com/Azure/azure-dev/actions/runs/4057804310/jobs/6983992673#step:5:8)3:1[8](https://github.com/Azure/azure-dev/actions/runs/4057804310/jobs/6983992673#step:5:9) - Unknown word (preinfradelete)
> ./schemas/v1.0/azure.yaml.json:278:18 - Unknown word (postinfradelete)
> ./schemas/v1.0/azure.yaml.json:283:18 - Unknown word (predown)
> ./schemas/v1.0/azure.yaml.json:288:18 - Unknown word (postdown)
> ./schemas/v1.0/azure.yaml.json:2[9](https://github.com/Azure/azure-dev/actions/runs/4057804310/jobs/6983992673#step:5:10)3:18 - Unknown word (preup)
> ./schemas/v1.0/azure.yaml.json:298:18 - Unknown word (postup)
> ./schemas/v1.0/azure.yaml.json:308:18 - Unknown word (postdeploy)
> ./schemas/v1.0/azure.yaml.json:3[13](https://github.com/Azure/azure-dev/actions/runs/4057804310/jobs/6983992673#step:5:14):[18](https://github.com/Azure/azure-dev/actions/runs/4057804310/jobs/6983992673#step:5:19) - Unknown word (prerestore)
> ./schemas/v1.0/azure.yaml.json:318:18 - Unknown word (postrestore)
> ```

@wbreza can you take a look?

**Review by wbreza — APPROVED** [review]
LGTM - Looks like their are some failing pipelines, but once those have been resolved this should be good to go.
Once you `rebase` the checks below should just go `green`.

**Comment by pamelafox** [comment]
(Sigh accidentally clicked close, re-opened now!)

File: templates/common/.devcontainer/devcontainer.json/python/devcontainer.json
Comment (jongio) [inline templates/common/.devcontainer/devcontainer.json/python/devcontainer.json:21]: We could leave as os-provided as that is the default and will be faster.

Comment (jongio) [inline templates/common/.devcontainer/devcontainer.json/python/devcontainer.json:21]: Can you change the container to base on:

mcr.microsoft.com/devcontainers/python:3.9

Instead of base, to see if it is smaller/faster?

Comment (pamelafox) [reply templates/common/.devcontainer/devcontainer.json/python/devcontainer.json:21]: Just want to clarify the idea:

If we change the container to "mcr.microsoft.com/devcontainers/python:3.9", then do we actually still want the Python feature? It doesn't seem like that's the intention of the feature - see my discussion with their team at https://github.com/devcontainers/features/issues/430 where I was trying to understand "os-provided".

Also, if we change the container to "mcr.microsoft.com/devcontainers/python:3.9", I presume we'd make the "python:3.9" an ARG to get overriden in devcontainer.json, for easier customization of each template?

I'm experimenting with that base of 3.9 now.

Comment (jongio) [reply templates/common/.devcontainer/devcontainer.json/python/devcontainer.json:21]: Yeah, if python base then you wouldn't need the feature.  Features are always slower, so might make sense, but wanted to see if faster, smaller if we do so.

Comment (pamelafox) [reply templates/common/.devcontainer/devcontainer.json/python/devcontainer.json:21]: Okay, here are the stats I found on my local M1.

I compared 
`FROM mcr.microsoft.com/devcontainers/python:3.9`
to
`FROM mcr.microsoft.com/vscode/devcontainers/base:0-bullseye`

|                    | python:3.9| base:0-bullseye + python os-provided | + python 3.9 |
|--------------------|-----------|-------------|------------|
| Base image size    | 1.29 GB    | 790.8 MB    | (same) |
| Dev Container size | 1.29 GB    | 1.12 GB     | 1.62 GB | 
| Time to build      | 12282 ms   | 59914 ms    | 94643 ms |

Observations:

* The Python 3.9 image size starts off larger than the bullseye image. The bulls eye + os-provided image is still smaller, but takes much more time to build. It doesn't seem safe to assume the OS has python, however, so I compared to python 3.9 feature, and then the image *and* build time are larger.
* The python image is really fast to build, comparatively!
* I tried adding on extensions to python 3.9 base image, and it didn't seem to change that size of the dev container image, so I think it installs them outside the image? Or maybe it wasn't affected because I already had those extensions locally. So I'm not 100% sure on effect of extensions, there are too many variables there.

So if we're going for faster loads, using a base python 3.9 image seems the way to go. That's a larger change though, since that means changing the other languages too, and setting it up so that an ARG changes the whole base image, not just the tag.

**Comment by jongio** [comment]
Yeah, thanks.

File: templates/common/.devcontainer/devcontainer.json/python/devcontainer.json
Comment (pamelafox) [reply templates/common/.devcontainer/devcontainer.json/python/devcontainer.json]: I've changed all the devcontainers to use the language-specific base image and not the feature. I've tested that they each successfully load (i.e. no issues creating the dev container) by trying a sample in each language with these changes. I haven't successfully gotten VS Code "Debug" working in every language, but I also don't know if "Debug" worked before my changes.

**Comment by pamelafox** [comment]
@jongio Updated to 3.10 and updated images to be language-specific (across all the languages), per my comment.

File: templates/common/.devcontainer/devcontainer.json/nodejs/devcontainer.json
Comment (pamelafox) [inline templates/common/.devcontainer/devcontainer.json/nodejs/devcontainer.json:6]: There's also a typescript image, I'm not sure if we'd prefer to use that one.

https://github.com/devcontainers/images/tree/main/src/typescript-node

vs

https://github.com/devcontainers/images/tree/main/src/javascript-node

File: templates/common/.devcontainer/devcontainer.json/python/devcontainer.json
Comment (pamelafox) [reply templates/common/.devcontainer/devcontainer.json/python/devcontainer.json]: Actually, it seems like I have gotten "Debug" working in them, at least to the point of it prompting to create a new azd environment. So it looks like the necessary debuggers are installed correctly.

**Comment by vhvb1989** [comment]
@jongio ping

File: templates/common/.devcontainer/devcontainer.json/java/devcontainer.json
Comment (vhvb1989) [inline templates/common/.devcontainer/devcontainer.json/java/devcontainer.json:6]: @pamelafox , looks like `Maven` is missing

We are getting:

```
Deploying apps for azd-template-test-todo-java-mongo-2644080-1...


ERROR: required external tools are missing:
 - error checking for external tool Maven: maven could not be found. Install either Maven or Maven Wrapper by visiting https://maven.apache.org/ or https://maven.apache.org/wrapper/

### ERROR: Dev container exec failed (exit code: 1)

```

Any idea how to fix it?  or should we revert the change for Java for now?

**Comment by vhvb1989** [comment]
This is currently blocked by: https://github.com/Azure/azure-dev/issues/1237

Deploying with pyhon is falling on CI due to symlink.  Should get unblock after: https://github.com/Azure/azure-dev/pull/1764

**Pull request merged**
